Mathlib/Data/Real/EReal.lean

```diff
@@ -1309,6 +1309,34 @@ theorem le_iff_le_forall_real_gt (x y : EReal) : (∀ z : ℝ, x < z → y ≤ z
       exact not_le_of_lt (coe_lt_top (_ + 1)) h
   · exact le_top
 
+theorem ge_iff_le_forall_real_lt (x y : EReal) : (∀ z : ℝ, z < y → z ≤ x) ↔ y ≤ x := by
+  symm
+  refine ⟨fun  h z z_lt_y ↦ le_trans (le_of_lt z_lt_y) h , ?_⟩
+  intro h
+  induction' x using EReal.rec with x
+  · apply le_of_eq ((eq_bot_iff_forall_lt y).2 _)
+    intro z
+    apply lt_of_not_le
+    intro z_le_y
+    apply not_le_of_lt (bot_lt_coe (z - 1))
+    specialize h (z-1)
+    apply h (lt_of_lt_of_le _ z_le_y)
+    norm_cast
+    exact sub_one_lt z
+  · induction' y using EReal.rec with y
+    · exact bot_le
+    · norm_cast
+      norm_cast at h
+      by_contra x_lt_y
+      rcases exists_between (lt_of_not_le x_lt_y) with ⟨z, ⟨x_lt_z, z_lt_y⟩⟩
+      specialize h z z_lt_y
+      exact not_le_of_lt x_lt_z h
+    · exfalso
+      specialize h (x + 1) (coe_lt_top (x+1))
+      norm_cast at h
+      exact not_le_of_lt (lt_add_one x) h
+  · exact le_top
```

Comment:
Let me explain the vast changes I'm suggesting here:
1. you can just use refine and swap the arguments of `⟨⟩` instead of using `symm`.
2. when using refine, you can introduce `?_` wherever, like in `fun h ↦ ?_`, and then you don't have to `intro` so many things
3. the `induction` syntax allows you to name the cases, which can be helpful, especially when you would like to name variables like `x y` in the second `h_real` case.
4. you can just inline your `specialize` bits where you use them.
5. `exact_mod_cast` can do `norm_cast` while supplying the term
6. when you use `by_contra!`, it uses `push_neg` on the introduced hypothesis.
```suggestion
  refine ⟨fun h ↦ ?_, fun h z x_lt_z ↦ le_trans h (le_of_lt x_lt_z)⟩
  induction x using EReal.rec
  case h_bot =>
    refine ((eq_bot_iff_forall_lt y).2 fun z ↦ ?_).le
    apply lt_of_le_of_lt <| h (z - 1) (bot_lt_coe (z - 1))
    exact_mod_cast sub_one_lt z
  case h_real =>
    induction y using EReal.rec
    case h_bot => exact bot_le
    case h_real x y =>
      norm_cast at h ⊢
      by_contra! x_lt_y
      rcases exists_between x_lt_y with ⟨z, x_lt_z, z_lt_y⟩
      exact not_le_of_lt z_lt_y <| h z x_lt_z
    case h_top =>
      exfalso
      exact not_le_of_lt (coe_lt_top (_ + 1)) <| h (_ + 1) (EReal.coe_lt_coe_iff.2 (lt_add_one _))
  case h_top => exact le_top
```

Reply:
That's amazing! Thank you very much for the explainer!